axelor-contract/src/main/java/com/axelor/apps/contract/service/ContractLineServiceImpl.java

```diff
@@ -382,27 +382,40 @@ public String computeProductDomain(Contract contract) {
     if (contract == null) {
       return domain;
     }
+    Company company = contract.getCompany();
     if (appBaseService.getAppBase().getEnableTradingNamesManagement()
         && appSaleService.getAppSale().getEnableSalesProductByTradName()
         && contract.getTradingName() != null
-        && contract.getCompany() != null
-        && !CollectionUtils.isEmpty(contract.getCompany().getTradingNameList())) {
+        && company != null
+        && !CollectionUtils.isEmpty(company.getTradingNameList())) {
       domain +=
           " AND " + contract.getTradingName().getId() + " member of self.tradingNameSellerSet";
     }
 
     int targetTypeSelect = contract.getTargetTypeSelect();
     if (targetTypeSelect == ContractRepository.CUSTOMER_CONTRACT
         || targetTypeSelect == ContractRepository.YEB_CUSTOMER_CONTRACT) {
-      domain += " AND self.sellable = true";
+      domain += buildDomain(company, "sellable");
     } else if (targetTypeSelect == ContractRepository.SUPPLIER_CONTRACT
         || targetTypeSelect == ContractRepository.YEB_SUPPLIER_CONTRACT) {
-      domain += " AND self.purchasable = true";
+      domain += buildDomain(company, "purchasable");
     }
 
     return domain;
   }
 
+  protected String buildDomain(Company company, String fieldName) {
+    if (company != null && productCompanyService.isCompanySpecificProductFields(fieldName)) {
+      return " AND EXISTS (SELECT pc FROM self.productCompanyList pc"
+          + " WHERE pc.company = "
+          + company.getId()
```

Comment:
With AOP 8.0,  comparing an object with an id won't work anymore. We should get used to it and start comparing object directly now to avoid migration issue. (For example, here, comparing Company with Company or Long with Long).